cpp/src/arrow/compute/exec/parsing.cc

```diff
@@ -0,0 +1,311 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cctype>
+#include <charconv>
+#include <string_view>
+#include <type_traits>
+#include <unordered_map>
+
+#include "arrow/compute/exec/expression.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace compute {
+static void ConsumeWhitespace(std::string_view& view) {
```

Comment:
Let's put everything inside an anonymous namespace instead of using static.

Reply:
I don't like anonymous namespaces because they are extremely debugger-hostile. I actually have an open issue to either disable anonymous namespaces in debug mode or to switch everything to using `static` everywhere.

Reply:
Until that goes through though I think we will have to stick to the current guidelines.

Reply:
Is this a guideline we currently have? The Google guide says:
```
All declarations can be given internal linkage by placing them in unnamed namespaces. Functions and variables can also be given internal linkage by declaring them static
```

So it seems that either way works according to the current guidelines, in which case I'd prefer to maintain compatibility with debuggers.

Reply:
You can make it `static` if you prefer. Generally we use the anonymous namespace since it's easy to forget the `static` keyword in front of a helper, and it doesn't work for non-functions.

Reply:
By the way this should be `void ConsumeWhitespace(std::string_view* view)` since the parameter is mutable.

Reply:
Sounds good, thanks for the clarification Antoine. I've changed it to take a pointer.